# Profitable App Profiles for the App Store and Google Play Markets

## About the project

This is a portfolio project where I act as a data analyst for an imaginary company that builds Android and iOS mobile apps for English-speaking markets. The apps are free to download and install on AppStore and Google Play. The main source of revenue for the company consists of in-app ads, which means that the more users download and install the app, the bigger the chance that they'll see and engage with in-app ads.  

## Project Goal

The goal of the project is to analyze data to help the developers understand what type of apps are likely to attract more users. 

## Datasets

To do this, I'll need to collect and analyze data about mobile apps available on Google Play and the App Store. 

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play. 

Collecting data for over 4 million apps requires a significant amount of time and money, so I'll try to analyze a sample of the data instead. To avoid spending resources on collecting new data by myself, I'll first try to see if I can find any relevant existing data at no cost. 

Luckily, there is [a learning dataset of AppStore apps available on Kaggle](tab:https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps). There is also a [public dataset of Google Play apps](tab:https://www.kaggle.com/datasets/lava18/google-play-store-apps). 

## Exploratory Data Analysis

Let's open both data sets and save them as lists of lists: 

In [4]:
from csv import reader

### The Google Play data set ###
opened_file = open('C:/Users/Olga Vogel/Desktop/googleplaystore.csv', encoding='utf8')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

### The App Store data set ###
opened_file = open('C:/Users/Olga Vogel/Desktop/AppleStore.csv', encoding='utf8')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

To make it easier to explore data, let's introduce an ```explore_data()``` function. 

The ```explore_data()``` function does the following:

* Takes in four parameters:
    * ```dataset```, which will be a list of lists
    * ```start``` and ```end```, which will both be integers and represent the starting and the ending indices of a slice from the dataset
    * ```rows_and_columns```, which will be a Boolean and has ```False``` as a default argument
* Slices the dataset using ```dataset[start:end]```
* Loops through the slice, and for each iteration, prints a row and adds a new line after that row using ```print('\n')```
    * The \n in print('\n') is a special character that won't print. Instead, the \n character adds a new line, and we use print('\n') to add some blank space between rows
* Prints the number of rows and columns if rows_and_columns is ```True```
    * dataset shouldn't have a header row, or the function will print the wrong number of rows (one more row compared to the actual length)

Let's also make the function show the number of rows and columns for any data set.

In [5]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [6]:
print(android_header)
print('\n')
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [7]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


The **Google Play** data set has **10841** apps and **13** columns. It seems that the columns that might be useful for future analysis are ```'App'```, ```'Category'```, ```'Reviews'```, ```'Installs'```, ```'Type'```, ```'Price'```, and ```'Genres'```.


The **AppStore** data set has **7197** iOS apps and **16** columns. The columns that seem interesting are: ```'track_name'```, ```'currency'```, ```'price'```, ```'rating_count_tot'```, ```'rating_count_ver'```, and ```'prime_genre'```. Not all column names are self-explanatory in this case, but details about each column can be found in the [data set documentation](tab:https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps).

## Data Cleaning

Before starting the analysis, I'll need to perform data cleaning, which includes removing or correcting wrong data, removing duplicate data, and modifying the data to fit the purpose of our analysis.

### Checking the data for empty strings and null values

Let's check if the data is consistent. 

In [8]:
android_header = android[0]
print(android[0],'\n')
print(len(android_header))
for row in android[1:]:
    if len(row) != len(android_header):
        print(row)
        print("\n")
        print("Index position is:", android.index(row))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

13
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Index position is: 10472


In [9]:
ios_header = ios[0]
print(ios[0],'\n')
print(len(ios_header))
for row in ios[1:]:
    if len(row) != len(ios_header):
        print(row)
        print("\n")
        print("Index position is:", ios.index(row))

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

16


The data set for AppStore seems OK, but there is a problem with line ```10472``` in Android data set, most likely a column shift has happened due to a missing value in the ```'Category'``` column. Let's once more compare this row against the header and another row that is correct. 

In [10]:
print(android[10472])  # incorrect row
print('\n')
print(android_header)  # header
print('\n')
print(android[0])      # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


The row ```10472``` corresponds to the app ```Life Made WI-Fi Touchscreen Photo Frame```, and we can see that the rating is ```19```. This is clearly off because the maximum rating for a Google Play app is ```5```. Indeed, a column shift has happened. Let's delete this row by using the [del statement](tab:https://docs.python.org/3/reference/simple_stmts.html?highlight=del#the-del-statement). 

In [11]:
print(len(android))
del android[10472]  # don't run this more than once
print(len(android))

10841
10840


### Removing Duplicate Statements

Let's check the data for duplicates. If we explore the Google Play data set long enough, we'll find that some apps have more than one entry. The same fact is being noticed in [the discussion section](tab:https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion) of the data set. 

The duplicates in this data set look like multiple data entries for one application - for example, Instagram: 

In [12]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Let's find out how many duplicate data entries there are in the data set. For this, we'll use a piece of code that counts the duplicate data entries by the name of the app: 

In [13]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
    
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We wouldn't want to count certain apps more than once for this business task. Let's see more examples and try to find out what's the difference between these data entries. 

In [14]:
for app in android:
    name = app[0]
    if name == 'Google My Business':
        print(app)

['Google My Business', 'BUSINESS', '4.4', '70991', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 24, 2018', '2.19.0.204537701', '4.4 and up']
['Google My Business', 'BUSINESS', '4.4', '70991', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 24, 2018', '2.19.0.204537701', '4.4 and up']
['Google My Business', 'BUSINESS', '4.4', '70991', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 24, 2018', '2.19.0.204537701', '4.4 and up']


In [15]:
for app in android:
    name = app[0]
    if name == 'Slack':
        print(app)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


In [16]:
for app in android:
    name = app[0]
    if name == 'HipChat - Chat Built for Teams':
        print(app)

['HipChat - Chat Built for Teams', 'BUSINESS', '3.8', '5868', '20M', '500,000+', 'Free', '0', 'Everyone', 'Business', 'July 3, 2018', '3.19.005', '4.1 and up']
['HipChat - Chat Built for Teams', 'BUSINESS', '3.8', '5868', '20M', '500,000+', 'Free', '0', 'Everyone', 'Business', 'July 3, 2018', '3.19.005', '4.1 and up']


In the previous example with ```'Instagram'``` and in the example with ```'Slack'```, there is a difference that happens on the fourth position of each row, which corresponds to the number of reviews. Probably, the data was collected at different times, which may explain the difference in the number of reviews. 


Let's suppose that the higher the number of reviews, the more reliable the ratings. Using this criterion, we can keep only the rows with the highest number of reviews, which will work well even with the cases like ```'Google My Business'```, where the rows are identical. 


To do that, we will:
* Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
* Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)


Let's find out the number of rows that we'll be left with when we remove the duplicates first. We already know that the number of duplicate rows is 1181. 

In [17]:
print('Expected length:', len(android) - 1181)

Expected length: 9659


Now, let's build a dictionary for unique apps. I will use ```in``` and ```not in``` operators to check for membership in the dictionary. 


To do this, I'll loop through the Google Play data set (excluding the header row). For each iteration, I'll do the following:
* Assign the app name to a variable named ```name```.
* Convert the number of reviews to ```float```. Assign it to a variable named ```n_reviews```.
* If ```name``` already exists as a key in the ```reviews_max``` dictionary and ```reviews_max[name] < n_reviews```, update the number of reviews for that entry in the ```reviews_max``` dictionary.
* If ```name``` is not in the ```reviews_max``` dictionary as a key, create a new entry in the dictionary where the key is the app name, and the value is the number of reviews. 

In [18]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Let's measure the length of the dictionary. 

In [19]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Everything seems to be fine. 

Now, let's use the ```reviews_max``` dictionary to remove the duplicates. For the duplicate cases, I'll only keep the entries with the highest number of reviews. In the code cell below:
* Start by initializing two empty lists, ```android_clean``` and ```already_added```.
* Loop through the Android data set, and for every iteration:
    * Isolate the name of the app and the number of reviews.
    * Add the current row (app) to the ```android_clean``` list, and the app name (name) to the ```already_added``` list if:
        * The number of reviews of the current app matches the number of reviews of that app as described in the ```reviews_max``` dictionary; and
        * The name of the app is not already in the ```already_added``` list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the ```'Google My Business'``` app has three entries, and the number of reviews is the same). If we just check for ```reviews_max[name] == n_reviews```, we'll still end up with duplicate entries for some apps.

In [20]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name) 

Let's confirm that the number of rows is as expected: 

In [21]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


We have ```9659``` rows, as expected. 

### Removing the data for apps not suitable for English-speaking markets

If we explore the data further, we'll find the apps that are possibly not suited for Enlgish-speaking markets, judging from their names. For example: 

In [22]:
print(ios[813][1])
print(ios[6731][1])

print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


We know that the characters that are specific to English texts are encoded using the ASCII standard, and each ASCII character has a corresponding number between 0 and 127 associated with it. We can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters. 


However, some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function that only checks for non-ASCII characters. 


To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters: 

In [23]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('لعبة تقدر تربح DZ'))
print(is_english('脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜'))

True
True
False
False
False


The function seems to work well. It's not perfect, and a few of non-English apps might still be present in the data. But for now, let's go with it and filter out the non-English apps for both data sets:

In [24]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)
        
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

We're left with ```9614```Android apps and ```6183``` iOS apps. 

### Removing the data for non-free apps

By the conditions of the task, the imaginary company is only interested in publishing the apps that are free to install and download. However, there are paid apps in both datasets. 

Let's remove the data on them using the price as the criterion. To do that, I'll identify the columns containing data on the prices and use strings ```'0'``` and ```'0.0'``` to filter out the free apps. 

In [25]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
print(len(android_final))
print(len(ios_final))

8864
3222


After doing this final step of data cleaning, we're left with ```8864``` Android apps and ```3222``` iOS apps. Let's get into business data analysis. 

## Business Data Analysis: Most Common Apps by Genre / Category

Our goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue.

To minimize risks and overhead, our validation strategy for an app idea has three steps:
* Build a minimal Android version of the app, and add it to Google Play.
* If the app has a good response from users, we develop it further.
* If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by determining the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our datasets.

### Building Frequency Tables

We'll need to build a frequency table for the ```prime_genre``` column of the App Store data set, and for the ```Genres``` and ```Category``` columns of the Google Play data set.

We'll build two functions we can use to analyze the frequency tables:
* One function to **generate frequency tables that show percentages**
* Another function we can use to **display the percentages in a descending order**. We'll need this, because dictionaries don't have order, and it will be very difficult to analyze the frequency tables without any order present. To do that,  we'll need to make use of the built-in ```sorted()``` function. This function takes in an iterable data type (like a list, dictionary, tuple, etc.), and returns a list of the elements of that iterable sorted in ascending or descending order (the reverse parameter controls whether the order is ascending or descending). It does not work well with dictionaries because it only considers and returns the dictionary keys. To use it properly, we'll need to transform the dictionary into a list of tuples, where each tuple contains a dictionary key along with its corresponding dictionary value. 

In [26]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### iOS Apps by Genre

Let's start by examining the frequency table for the ```prime_genre``` column of the App Store data set.

In [27]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Free-to-download English-speaking market apps in App Store are mostly what can be considered games (58.16%). Around 8% of apps are entertainment apps, followed by photo and video apps (around 5%). Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

Observing this statistics, it's safe to conclude that App Store (at least the part containing free apps in English) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.).  Apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare.

However, the fact that gaming apps are the most numerous does not imply that they also have the greatest number of users. We're only observing the relative popularity of the genre among app publishers. 

### Android Apps by Genre

In [28]:
display_table(android_final, 1) # Category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The analysis of ```Category``` column of Google Play datasets shows that a category called FAMILY accounts for almost 19% of the apps. If we look into this category directly on Google Play, we'll see that it mostly contains games for kids. However, the general representation of non-game categories on Google Play seems slightly better than in AppStore: the TOOLS, LIFESTYLE, PRODUCTIVITY, FINANCE, MEDICAL etc.. categories certainly have their own visible part on the market.

Let's observe the ```Genres``` column: 

In [29]:
display_table(android_final, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

This column kind of supports the idea that non-game apps are better present on Google Play. However, the classification itself seems more granular, with more of what looks like sub-categories. For the sake of having bird's-eye view, let's use ```Category``` column instead. 

### Preliminary Conclusions

For now, we can conclude that App Store is dominated by apps designed for entertainment, while Google Play shows a more balanced landscape of both practical and entertainment apps. However, this data is mostly **helpful in learning how app publishers behave**. Let's additionally investigate the user dynamics and see **which kinds of apps have the biggest number of users**.

### App Store: Most Popular Apps by Genre 

Let's start with iOS apps. 

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the ```Installs``` column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the ```rating_count_tot``` app.

Let's start with calculating the average number of user ratings per app genre on the App Store. To do that, we'll need to do the following:
* Isolate the apps of each genre
* Add up the user ratings for the apps of that genre
* Divide the sum by the number of apps belonging to that genre (not by the total number of apps)

To do this, we'll need to use *nested loop*:  
* Generate a frequency table for the ```prime_genre``` column to get the unique app genres (below, we'll need to loop over the unique genres). We'll use the ```freq_table()``` function we wrote before.
* Loop over the unique genres of the App Store dataset. For each iteration:
  * Initiate a variable named ```total``` with a value of ```0```. This variable will store the sum of user ratings (the number of ratings, not the actual ratings) specific to each genre.
  * Initiate a variable named ```len_genre``` with a value of ```0```. This variable will store the number of apps specific to each genre.
   * Loop over the App Store dataset, and for each iteration, do the following:
        * Save the app genre to a variable named ```genre_app```.
        * If ```genre_app``` is the same as ```genre``` (the iteration variable of the main loop), then do the following:
            * Save the number of user ratings of the app as a float.
            * Add up the number of user ratings to the total variable.
            * Increment the ```len_genre``` variable by 1.
    * Compute the average number of user ratings by dividing ```total``` by ```len_genre```. 
    * Print the app genre and the average number of user ratings. 

In [38]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


We can observe that some app genres have the biggest number of reviews (Navigation, Social Networking, Music etc..). However, if we take, for example, Social Networking and Navigation categories, we'll see the following: 

In [31]:
for app in ios_final:
    if app[-5] == 'Social Networking':
        print(app[1], ':', app[5]) # name and number of ratings

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

In [36]:
for app in ios_final:
    if app[-5] == 'Music':
        print(app[1], ':', app[5]) # name and number of ratings

Pandora - Music & Radio : 1126879
Spotify Music : 878563
Shazam - Discover music, artists, videos & lyrics : 402925
iHeartRadio – Free Music & Radio Stations : 293228
SoundCloud - Music & Audio : 135744
Magic Piano by Smule : 131695
Smule Sing! : 119316
TuneIn Radio - MLB NBA Audiobooks Podcasts Music : 110420
Amazon Music : 106235
SoundHound Song Search & Music Player : 82602
Sonos Controller : 48905
Bandsintown Concerts : 30845
Karaoke - Sing Karaoke, Unlimited Songs! : 28606
My Mixtapez Music : 26286
Sing Karaoke Songs Unlimited with StarMaker : 26227
Ringtones for iPhone & Ringtone Maker : 25403
Musi - Unlimited Music For YouTube : 25193
AutoRap by Smule : 18202
Spinrilla - Mixtapes For Free : 15053
Napster - Top Music & Radio : 14268
edjing Mix:DJ turntable to remix and scratch music : 13580
Free Music - MP3 Streamer & Playlist Manager Pro : 13443
Free Piano app by Yokee : 13016
Google Play Music : 10118
Certified Mixtapes - Hip Hop Albums & Mixtapes : 9975
TIDAL : 7398
YouTube Mu

These categories are populated by such giants as Facebook, Pinterest, Skype, Pandora, Spotify, Waze, Google Maps etc.., which obviously invest in user acquisition and are hard to compete against. For now, let's look at other categories instead. 

In [33]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


In [35]:
for app in ios_final:
    if app[-5] == 'Book':
        print(app[1], ':', app[5])

Kindle – Read eBooks, Magazines & Textbooks : 252076
Audible – audio books, original series & podcasts : 105274
Color Therapy Adult Coloring Book for Adults : 84062
OverDrive – Library eBooks and Audiobooks : 65450
HOOKED - Chat Stories : 47829
BookShout: Read eBooks & Track Your Reading Goals : 879
Dr. Seuss Treasury — 50 best kids books : 451
Green Riding Hood : 392
Weirdwood Manor : 197
MangaZERO - comic reader : 9
ikouhoushi : 0
MangaTiara - love comic reader : 0
謎解き : 0
謎解き2016 : 0


Both ```Book``` and ```Reference``` categories seem interesting. Apart from them being populated by Kindle and Audible, there is also a phenomenon of making an app out of a really popular book, like Bible, a dictionary or even a coloring book for adults. It seems that we might have some success if we find another popular book to make into an app. 

Another strategy would be to suggest additional features or alternative payment model to a book app to make it more interesting: 
* gamification features, such as a "reading strike" or achievements that promote daily reading
* in-app dictionaries to attract users who otherwise download a separate dictionary app
* a read-aloud AI feature to accomodate the needs of people with kids
* a way to buy books one by one or lend an e-book much like you do from the library

If we can combine a previously "pure entertainment" content app with some lifestyle & productivity features, this app will have a chance to stand out on the market populated with pure for-fun apps. 

Other genres that seem popular include weather, food and drink and finance apps. It will be extremely difficult to quickly launch a lovable food&drink app without a certain connection to a well-known restaurant - or to launch a cooking&delivery business from scratch. The same idea stops us from seriously considering the finance sphere. As for weather apps, the time spent in such an app is merely seconds. The chances of making profit from in-app adds are extremely low. 

That being said, let's stick with "bookish lifestyle" idea for a bit more and observe what happens on the Google Play market. 

### Google Play: Most Popular Apps by Genre

For the Google Play market, we actually have data about the number of installs in the ```Installs``` column, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [37]:
display_table(android_final, 5) 

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to find out which app genres attract the most users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, however, we'll need to convert each install number from a string to a float. This means we need to remove the commas and the plus characters, or the conversion will fail and cause an error.

To remove characters from strings, we can use the ```str.replace(old, new)``` method. ```str.replace()``` takes in two parameters, ```old``` and ```new```, and replaces all occurrences of old within a string with new. To remove certain characters, we can replace them with the empty string ```''```.

Let's calculate the verage number of installs per app genre for the Google Play dataset. As in the previous case, we'll also use the *nested loop*: 
* Generate a frequency table for the ```Category``` column of the Google Play dataset to get the unique app genres. We'll use the ```freq_table()``` function once more.
* Loop over the unique genres of the Google Play dataset. For each iteration, do the following:
    * Initiate a variable named ```total``` with a value of ```0```. This variable will store the sum of installs specific to each genre.
    * Initiate a variable named ```len_category``` with a value of ```0```. This variable will store the number of apps specific to each genre.
    * Loop over the Google Play dataset, and for each iteration, do the following:
        * Save the app genre to a variable named ```category_app```.
        * If ```category_app``` is the same as ```category``` (the iteration variable of the main loop), then do the following:
            * Save the number of installs.
            * Remove any ```+``` or ```,``` character, and then convert the string to a float.
            * Add up the number of installs to the ```total``` variable.
            * Increment the ```len_category``` variable by 1.
    * Compute the average number of installs by dividing ```total``` by ```len_category```. 
    * Print the app genre and the average number of installs. 

In [39]:
categories_android = freq_table(android_final, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [40]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

We see the same pattern for the video players category, where the market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.). Again, the main concern is that these niches seem to be dominated by a few giants who are hard to compete against.

Let's have a look at the ```BOOKS_AND_REFERENCE``` category which showed some potential among App Store apps before.

In [41]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a huge variety of apps: readers, dictionaries, popular books like Bible and Al Quran, programming tutorials and specific projects like SCP Foundation. It is very obvious that there also are some giants that populate the market, Google Play Books and Kindle among them. 



In [42]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

If we observe the level of medium popularity, we'll find pretty much the same situation as in App Store: building an app around a popular book OR adding supplementary functionality to a reader app, turning it into a productivity & lifestyle app seems to be a nice idea.

## Conclusions

This was a portfolio project where I acted as a data analyst for an imaginary company that builds Android and iOS mobile apps for English-speaking markets. The apps are free to download and install on AppStore and Google Play. The main source of revenue for the company consists of in-app ads, which means that the more users download and install the app, the bigger the chance that they'll see and engage with in-app ads. 

The goal was to recommend an app profile that can be profitable for both markets.

I performed exploratory data analysis on two data sets: [a learning dataset of AppStore apps available on Kaggle](tab:https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps) and a [public dataset of Google Play apps](tab:https://www.kaggle.com/datasets/lava18/google-play-store-apps). 

After performing data cleaning (which includes removing or correcting wrong data, removing duplicate data, and modifying the data to fit the purpose of our analysis), I observed the data sets to learn which app genres are mostly present on both stores and whether they are popular with the users. 

I ended up with recommending a book app enhanced with additional functionalities, which falls into the category of popular apps and allows the company to effectively compete in two genres: books/reference and productivity/lifestyle. Possible product development includes: 
* gamification features, such as a "reading strike" or achievements that promote daily reading
* in-app dictionaries to attract users who otherwise download a separate dictionary app
* a read-aloud AI feature to accomodate the needs of people with kids and compete on the extra market (entertainment apps for kids) 
* alternative monetization models, such as: a way to buy books one by one; lending an e-book for a limited period of time; using internal currency; subscription-based models which are equally interesting for B2B business development opportunities